File: app/src/main/java/com/example/srtwatcher/domain/Train.kt
package com.example.srtwatcher.domain

import java.time.LocalTime

/**
 * DOM 에서 추출한 한 편성의 열차 정보. (DESIGN.md §7)
 *
 * 이 모델은 WebView / DOM selector 와 완전히 분리되어 있으며,
 * SelectionEngine 과 UI 는 이 모델만 알고 있으면 된다.
 */
data class Train(
    val trainNumber: String,
    val departureStation: String,
    val arrivalStation: String,
    val departureTime: LocalTime,
    val arrivalTime: LocalTime,
    val generalSeat: SeatStatus,
    val firstClassSeat: SeatStatus,
) {
    /** 재조회 후에도 같은 열차를 알아보기 위한 식별자. (사용자 선택을 유지하는 데 쓴다) */
    val key: TrainKey
        get() = TrainKey(trainNumber = trainNumber, departureTime = departureTime)

    fun seatStatusOf(seatClass: SeatClass): SeatStatus = when (seatClass) {
        SeatClass.GENERAL -> generalSeat
        SeatClass.FIRST_CLASS -> firstClassSeat
    }

    /** "SRT 339  수서 → 부산  18:30 → 21:05" */
    fun summary(): String = buildString {
        append(trainNumber)
        append("  ")
        append(departureStation)
        append(" → ")
        append(arrivalStation)
        append("  ")
        append(departureTime.toString())
        append(" → ")
        append(arrivalTime.toString())
    }
}

File: extension/src/domain/train.js
// DOM 에서 추출한 한 편성. (android: domain/Train.kt, DESIGN.md §7)
//
// 이 모델은 DOM selector 도 chrome API 도 모른다. SelectionEngine 과 화면은
// 이 모델만 알면 된다.
//
// 안드로이드의 `data class Train` 에 대응하지만 **평범한 객체**로 둔다.
// service worker → popup 으로 그대로 건너가야 하는 값이라, 메서드를 달면
// 경계를 넘는 순간 사라진다. 동작은 아래 함수들이 갖는다.

import { SeatClass } from './seat-class.js';
import { SeatStatus } from './seat-status.js';
import { trainKey } from './train-key.js';

/**
 * 재조회 후에도 같은 열차를 알아보기 위한 식별자. (§38-4)
 * 사용자 선택을 유지하는 데 쓴다.
 */
export function keyOf(train) {
  return trainKey(train.trainNumber, train.departureTime);
}

export function seatStatusOf(train, seatClass) {
  if (!train) return SeatStatus.UNKNOWN;
  return seatClass === SeatClass.GENERAL ? train.generalSeat : train.firstClassSeat;
}

/** "305  동탄 → 김천구미  07:11 → 08:17" */
export function trainSummary(train) {
  return `${train.trainNumber}  ${train.departureStation} → ${train.arrivalStation}  ` +
    `${train.departureTime} → ${train.arrivalTime}`;
}
